usr/sbin/p11sak/p11sak.c

```diff
@@ -2455,13 +2943,121 @@ static CK_RV p11sak_list_key(void)
     return CKR_OK;
 }
 
-static CK_RV p11sak_remove_key(void)
+static char prompt_user(const char *message, char* allowed_chars)
 {
-    // TODO
+    int len;
+    size_t linelen = 0;
+    char *line = NULL;
+    char ch = '\0';
+
+    printf("%s", message);
+
+    while (1) {
+        len = getline(&line, &linelen, stdin);
+        if (len == -1)
+            break;
+
+        if (strlen(line) == 2 && strpbrk(line, allowed_chars) != 0) {
+            ch = line[0];
+            break;
+        }
+
+        warnx("Improper reply, try again");
+    }
+
+    if (line != NULL)
+        free(line);
+
+    return ch;
+}
+
+static CK_RV handle_key_remove(CK_OBJECT_HANDLE key, CK_OBJECT_CLASS class,
+                               const struct p11sak_keytype *keytype,
+                               CK_ULONG keysize, const char *typestr,
+                               const char* label, void *private)
+{
+    struct p11sak_remove_data *data = private;
+    char *msg = NULL;
+    char ch;
+    CK_RV rc;
+
+    UNUSED(class);
+    UNUSED(keysize);
+    UNUSED(keytype);
+
+    if (data->skip_all) {
+        data->num_skipped++;
+        return CKR_OK;
+    }
+
+    if (!data->remove_all) {
+        if (asprintf(&msg, "Are you sure you want to remove %s key object \"%s\" [y/n/a/c]? ",
+                     typestr, label) < 0 ||
+            msg == NULL) {
+            warnx("Failed to allocate memory for a message");
+            return CKR_HOST_MEMORY;
+        }
+        ch = prompt_user(msg, "ynac");
+        free(msg);
+
+        switch (ch) {
+        case 'n':
+            data->num_skipped++;
+            return CKR_OK;
+        case 'c':
+            data->skip_all = true;
+            data->num_skipped++;
+            return CKR_OK;
+        case 'a':
+            data->remove_all = true;
+            break;
+        default:
+            break;
+        }
+    }
+
+    rc = pkcs11_funcs->C_DestroyObject(pkcs11_session, key);
+    if (rc != CKR_OK) {
+        warnx("Failed to remove %s key object \"%s\": C_DestroyObject: 0x%lX: %s",
+                typestr, label, rc, p11_get_ckr(rc));
+        data->num_failed++;
+        return CKR_OK;
+    }
+
+    printf("Successfully removed %s key object \"%s\".\n", typestr, label);
+    data->num_removed++;
 
     return CKR_OK;
 }
 
+static CK_RV p11sak_remove_key(void)
```

Comment:
Hmm ... we would really tend to p11sak_delete_key

Reply:
see above